src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -157,39 +162,182 @@ internal void ClearCaches()
+                {
+                    if (current == JsonTypeInfo.ObjectType)
+                    {
+                        // Avoid falling back to the contract for object since it's polymorphic
+                        // and it would try to send us back to the runtime type that isn't supported.
+                        break;
+                    }
+
+                    candidate = GetOrAddCacheEntry(current);
+                    if (candidate.HasResult)
+                    {
+                        // We found a type in the class hierarchy that has a contract -- stop looking further up.
+                        candidateType = current;
+                        break;
+                    }
+                }
+
+                foreach (Type interfaceType in type.GetInterfaces())
+                {
+                    CacheEntry interfaceEntry = GetOrAddCacheEntry(interfaceType);
+                    if (interfaceEntry.HasResult)
+                    {
+                        if (candidateType != null)
+                        {
+                            if (interfaceType.IsAssignableFrom(candidateType))
+                            {
+                                // The previous candidate is more derived than the
+                                // current interface -- keep our previous choice.
+                                continue;
+                            }
+                            else if (candidateType.IsAssignableFrom(interfaceType))
+                            {
+                                // The current interface is more derived than the
+                                // previous candidate -- replace the candidate value.
+                            }
+                            else
+                            {
+                                // We have found two possible ancestors that are not in subtype relationship.
+                                // This indicates we have encountered a diamond ambiguity -- abort search and record an exception.
+                                NotSupportedException nse = ThrowHelper.GetNotSupportedException_AmbiguousMetadataForType(type, candidateType, interfaceType);
+                                candidate = new CacheEntry(ExceptionDispatchInfo.Capture(nse));
+                                break;
+                            }
+                        }
+
+                        candidate = interfaceEntry;
+                        candidateType = interfaceType;
+                    }
+                }
+
+                entry.NearestAncestor = candidate;
+                entry.IsNearestAncestorResolved = true;
+                return candidate;
+            }
+
+            private sealed class CacheEntry
             {
-                _jsonTypeInfoCache.Clear();
+                public readonly bool HasResult;
+                public readonly JsonTypeInfo? TypeInfo;
+                public readonly ExceptionDispatchInfo? ExceptionDispatchInfo;
```

Comment:
nit should this be `private`? It's not called elsewhere.

Reply:
The class itself is private, but its members are accessed by the containing class.